File: lib-database/src/main/java/com/ustadmobile/core/db/dao/SaleProductPictureDao.java
package com.ustadmobile.core.db.dao;

import com.ustadmobile.core.db.UmLiveData;
import com.ustadmobile.core.impl.UmCallback;
import com.ustadmobile.lib.database.annotation.UmDao;
import com.ustadmobile.lib.database.annotation.UmDbGetAttachment;
import com.ustadmobile.lib.database.annotation.UmDbSetAttachment;
import com.ustadmobile.lib.database.annotation.UmQuery;
import com.ustadmobile.lib.database.annotation.UmRepository;
import com.ustadmobile.lib.database.annotation.UmRestAccessible;
import com.ustadmobile.lib.db.entities.SaleProductPicture;
import com.ustadmobile.lib.db.sync.dao.SyncableDao;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;

@UmDao(hasAttachment = true,
        permissionJoin = " LEFT JOIN SaleProduct ON SaleProductPicture.saleProductPictureSaleProductUid = SaleProduct.saleProductUid",
        selectPermissionCondition = RoleDao.SELECT_ACCOUNT_IS_ADMIN,
        updatePermissionCondition = RoleDao.SELECT_ACCOUNT_IS_ADMIN)
@UmRepository
public abstract class SaleProductPictureDao implements SyncableDao<SaleProductPicture, SaleProductPictureDao> {

    @UmQuery("SELECT * FROM SaleProductPicture WHERE SaleProductPicture.saleProductPictureUid = :uid")
    public abstract UmLiveData<SaleProductPicture> findByUidLive(long uid);


    @UmQuery("SELECT * FROM SaleProductPicture where saleProductPictureSaleProductUid = :uid ORDER BY " +
            " saleProductPictureTimestamp DESC LIMIT 1")
    public abstract UmLiveData<SaleProductPicture> findByProductUidLive(long uid);

    @UmDbSetAttachment
    public abstract void setAttachment(long uid, InputStream pictureAttachment) throws IOException;

    @UmDbGetAttachment
    public abstract String getAttachmentPath(long uid);

    @UmDbGetAttachment
    public abstract InputStream getAttachmentStream(long uid) throws IOException;

    @UmDbSetAttachment
    public abstract void setAttachmentFromTmpFile(long uid, File tmpFile);

    @UmQuery("SELECT * FROM SaleProductPicture where saleProductPictureSaleProductUid = :saleProductUid ORDER BY saleProductPictureTimestamp DESC LIMIT 1")
    public abstract void findBySaleProductUidAsync(long saleProductUid, UmCallback<SaleProductPicture> findByUidCallback);

    @UmDbSetAttachment
    @UmRestAccessible
    @UmRepository(delegateType = UmRepository.UmRepositoryMethodType.DELEGATE_TO_WEBSERVICE)
    public abstract void uploadAttachment(long uid, InputStream attachment) throws IOException;

    @UmDbGetAttachment
    @UmRestAccessible
    @UmRepository(delegateType = UmRepository.UmRepositoryMethodType.DELEGATE_TO_WEBSERVICE)
    public abstract InputStream downloadAttachment(long uid) throws IOException;
}

File: lib-database/src/commonMain/kotlin/com/ustadmobile/core/db/dao/SaleProductPictureDao.kt
package com.ustadmobile.core.db.dao

import androidx.paging.DataSource
import androidx.room.Dao
import androidx.room.Query
import com.ustadmobile.door.DoorLiveData
import com.ustadmobile.door.annotation.GetAttachmentData
import com.ustadmobile.door.annotation.SetAttachmentData
import com.ustadmobile.lib.database.annotation.UmDao
import com.ustadmobile.lib.database.annotation.UmRepository
import com.ustadmobile.lib.db.entities.SaleProductPicture

@UmDao(hasAttachment = true, permissionJoin = " LEFT JOIN SaleProduct ON SaleProductPicture.saleProductPictureSaleProductUid = SaleProduct.saleProductUid", selectPermissionCondition = RoleDao.SELECT_ACCOUNT_IS_ADMIN, updatePermissionCondition = RoleDao.SELECT_ACCOUNT_IS_ADMIN)
@UmRepository
@Dao
abstract class SaleProductPictureDao : BaseDao<SaleProductPicture> {


    @Query("SELECT * FROM SaleProductPicture WHERE SaleProductPicture.saleProductPictureUid = :uid")
    abstract fun findByUidLive(uid: Long): DoorLiveData<SaleProductPicture?>

    @Query("SELECT * FROM SaleProductPicture where saleProductPictureSaleProductUid = :uid ORDER BY " + " saleProductPictureTimestamp DESC LIMIT 1")
    abstract fun findByProductUidLive(uid: Long): DoorLiveData<SaleProductPicture?>

    @Query("""SELECT * FROM SaleProductPicture WHERE SaleProductPicture.saleProductPictureUid = :productPictureUid""")
    abstract fun findByUidAsync(productPictureUid: Long): SaleProductPicture?

    @SetAttachmentData
    open fun setAttachment(entity: SaleProductPicture, filePath: String) {
        throw Exception(Exception("Shouldn't call the Dao, call the repo instead "))
    }

    @GetAttachmentData
    open fun getAttachmentPath(entity: SaleProductPicture): String? {
        return ""
    }

    @Query(""" SELECT * FROM SaleProductPicture WHERE 
        saleProductPictureSaleProductUid = :saleProductUid AND saleProductPictureIndex = 
         ( SELECT MIN(saleProductPictureIndex) FROM SaleProductPicture WHERE saleProductPictureSaleProductUid = :saleProductUid ) """)
    abstract suspend fun findBySaleProductUidAsync2(saleProductUid: Long): SaleProductPicture?

    @Query("""SELECT * FROM SaleProductPicture WHERE 
        saleProductPictureSaleProductUid = :productUid ORDER BY saleProductPictureIndex ASC""")
    abstract suspend fun findAllByProductByIndexAsync(productUid: Long) : List<SaleProductPicture>

    @Query("""SELECT * FROM SaleProductPicture WHERE 
        saleProductPictureSaleProductUid = :productUid ORDER BY saleProductPictureIndex ASC""")
    abstract fun findAllByProductByIndex(productUid: Long): DataSource.Factory<Int,SaleProductPicture>

    @Query("""SELECT CASE WHEN MAX(saleProductPictureIndex) IS NOT NULL 
        THEN MAX(saleProductPictureIndex) ELSE -1 END FROM SaleProductPicture 
        WHERE saleProductPictureSaleProductUid = :productUid""")
    abstract suspend fun findMaxIndexForSaleProductPicture(productUid : Long): Int

    @Query("""SELECT COUNT(*) FROM SaleProductPicture 
        WHERE saleProductPictureSaleProductUid = :productUid 
        AND saleProductPictureIndex >=0 """)
    abstract suspend fun findTotalNumberOfPicturesForAProduct(productUid: Long): Int

    //TODO: Check logic if it is okay.

    @Query("""UPDATE SaleProductPicture 
        SET saleProductPictureIndex = saleProductPictureIndex + 1 
        , saleProductPictureLCB = (SELECT nodeClientId FROM SyncNode LIMIT 1)
        WHERE saleProductPictureIndex >= :index 
        AND saleProductPictureSaleProductUid = :productUid""")
    abstract suspend fun moveIndexAheadForSaleProductAndIndex(productUid: Long, index: Int): Int

    @Query("""UPDATE SaleProductPicture SET saleProductPictureIndex = :index 
        , saleProductPictureLCB = (SELECT nodeClientId FROM SyncNode LIMIT 1)
        WHERE SaleProductPictureUid = :productPictureUid """)
    abstract suspend fun updateSaleProductPictureIndex(productPictureUid: Long, index: Int): Int

    suspend fun changeIndexForSaleProductPicture(productPictureUid: Long, productUid: Long, index: Int){
        moveIndexAheadForSaleProductAndIndex(productUid, index)
        updateSaleProductPictureIndex(productPictureUid, index)
    }


}
